File: bindings/python/STRling/simply/sets.py
from .pattern import STRlingError, Pattern, lit


def between(start: str, end: str, min_rep: int = None, max_rep: int = None):
    """
    Matches all characters within and including the start and end of a letter or digit range.

    Parameters:
    - start (str or int): The starting character or digit of the range.
    - end (str or int): The ending character or digit of the range.
    - min_rep (optional): Specifies the minimum digit of characters to match.
    - max_rep (optional): Specifies the maximum digit of characters to match.

    Returns:
    - An instance of the Pattern class.

    Examples:
        ```
        # Matches any digit from 0 to 9.
        my_pattern1 = s.between(0, 9)

        # Matches any lowercase letter from 'a' to 'z'.
        my_pattern2 = s.between('a', 'z')

        # Matches any uppercase letter from 'A' to 'Z'.
        my_pattern3 = s.between('A', 'Z')
        ```
    """

    if not (isinstance(start, str) and isinstance(end, str)) and not (isinstance(start, int) and isinstance(end, int)):
        message = """
        Method: simply.between(start, end)

        The 'start' and 'end' arguments must both be integers (0-9) or letters of the same case (A-Z or a-z).
        """
        raise STRlingError(message)

    if isinstance(start, int):
        if start > end:
            message = """
            Method: simply.between(start, end)

            The 'start' integer must not be greater than the 'end' integer.
            """
            raise STRlingError(message)

        if not (0 <= start <= 9 and 0 <= end <= 9):
            message = """
            Method: simply.between(start, end)

            The 'start' and 'end' integers must be single digits (0-9).
            """
            raise STRlingError(message)

        new_pattern = f'[{start}-{end}]'

    if isinstance(start, str):
        if not start.isalpha() or not end.isalpha():
            message = """
            Method: simply.between(start, end)

            The 'start' and 'end' must be alphabetical characters.
            """
            raise STRlingError(message)

        if len(start) != 1 or len(end) != 1:
            message = """
            Method: simply.between(start, end)

            The 'start' and 'end' characters must be single letters.
            """
            raise STRlingError(message)

        if start.islower() != end.islower():
            message = """
            Method: simply.between(start, end)

            The 'start' and 'end' characters must be of the same case.
            """
            raise STRlingError(message)

        if start > end:
            message = """
            Method: simply.between(start, end)

            The 'start' character must not be lexicographically greater than the 'end' character.
            """
            raise STRlingError(message)

        new_pattern = f'[{start}-{end}]'

    return Pattern(new_pattern, custom_set=True)(min_rep, max_rep)


def not_between(start: str, end: str, min_rep: int = None, max_rep: int = None):
    """
    Matches any character not within or including the start and end of a letter or digit range.

    Parameters:
    - start (str or int): The starting character or digit of the range.
    - end (str or int): The ending character or digit of the range.
    - min_rep (optional): Specifies the minimum digit of characters to match.
    - max_rep (optional): Specifies the maximum digit of characters to match.

    Returns:
    - An instance of the Pattern class.

    Examples:
        ```
        # Matches any character that is not a digit from 0 to 9.
        my_pattern1 = s.not_between(0, 9)

        # Matches any character that is not a lowercase letter from 'a' to 'z'.
        my_pattern2 = s.not_between('a', 'z')

        # Matches any character that is not a uppercase letter from 'A' to 'Z'.
        my_pattern3 = s.not_between('A', 'Z')
        ```
    """

    if not (isinstance(start, str) and isinstance(end, str)) and not (isinstance(start, int) and isinstance(end, int)):
        message = """
        Method: simply.not_between(start, end)

        The 'start' and 'end' arguments must both be integers (0-9) or letters of the same case (A-Z or a-z).
        """
        raise STRlingError(message)

    if isinstance(start, int):
        if start > end:
            message = """
            Method: simply.not_between(start, end)

            The 'start' integer must not be greater than the 'end' integer.
            """
            raise STRlingError(message)

        if not (0 <= start <= 9 and 0 <= end <= 9):
            message = """
            Method: simply.not_between(start, end)

            The 'start' and 'end' integers must be single digits (0-9).
            """
            raise STRlingError(message)

        new_pattern = f'[^{start}-{end}]'

    if isinstance(start, str):
        if not start.isalpha() or not end.isalpha():
            message = """
            Method: simply.not_between(start, end)

            The 'start' and 'end' must be alphabetical characters.
            """
            raise STRlingError(message)

        if len(start) != 1 or len(end) != 1:
            message = """
            Method: simply.not_between(start, end)

            The 'start' and 'end' characters must be single letters.
            """
            raise STRlingError(message)

        if start.islower() != end.islower():
            message = """
            Method: simply.not_between(start, end)

            The 'start' and 'end' characters must be of the same case.
            """
            raise STRlingError(message)

        if start > end:
            message = """
            Method: simply.not_between(start, end)

            The 'start' character must not be lexicographically greater than the 'end' character.
            """
            raise STRlingError(message)

        new_pattern = f'[^{start}-{end}]'

    return Pattern(new_pattern, custom_set=True, negated=True)(min_rep, max_rep)


def in_chars(*patterns):
    """
    Matches any provided patterns, but they can't include subpatterns.

    Parameters:
    - patterns (Pattern/str): One or more non-composite patterns to match.

    Returns:
    - An instance of the Pattern class.

    Examples:
        ```
        # Matches any letter, digit, comma, and period.
        my_pattern = s.in_chars(s.letter(), s.digit(), ',.')
        ```
    """

    # Check all patterns are instance of Pattern or str
    clean_patterns = []
    for pattern in patterns:
        if isinstance(pattern, str):
            pattern = lit(pattern)

        if not isinstance(pattern, Pattern):
            message = """
            Method: simply.in_chars(*patterns)

            The parameters must be instances of `Pattern` or `str`.

            Use a string such as "123abc$" to match literal characters, or use a predefined set like `simply.letter()`.
            """
            raise STRlingError(message)

        clean_patterns.append(pattern)

    if any(p.composite for p in clean_patterns):
        message = """
        Method: simply.in_chars(*patterns)

        All patterns must be non-composite.
        """
        raise STRlingError(message)

    joined = r''
    for pattern in clean_patterns:
        if len(str(pattern)) > 1 and str(pattern)[-1] == '}' and str(pattern)[-2] != "\\":
            message = """
            Method: simply.in_chars(*patterns)

            Patterns must not have specified ranges.
            """
            raise STRlingError(message)

        if pattern.custom_set:
            if pattern.negated:
                message = """
                Method: simply.in_chars(*patterns)

                To match the characters specified in a negated set, move the parameters directly into simply.in_chars(*patterns).

                Example: simply.in_chars(simply.not_in_chars(*patterns)) => simply.in_chars(*patterns)
                """
                raise STRlingError(message)
            else:
                joined += str(pattern)[1:-1]  # [chars] => chars
        else:
            joined += str(pattern)

    new_pattern = f'[{joined}]'
    return Pattern(new_pattern, custom_set=True)

def not_in_chars(*patterns):
    """
    Matches anything but the provided patterns, but they can't include subpatterns.

    Parameters:
    - patterns (Pattern/str): One or more non-composite patterns to avoid.

    Returns:
    - An instance of the Pattern class.

    Examples:
        ```
        # Matches any character that is not a letter, digit, comma, and period.
        my_pattern = s.not_in_chars(s.letter(), s.digit(), ',.')
        ```
    """

    # Check all patterns are instance of Pattern or str
    clean_patterns = []
    for pattern in patterns:
        if isinstance(pattern, str):
            pattern = lit(pattern)

        if not isinstance(pattern, Pattern):
            message = """
            Method: simply.not_in_chars(*patterns)

            The parameters must be instances of `Pattern` or `str`.

            Use a string such as "123abc$" to match literal characters, or use a predefined set like `simply.letter()`.
            """
            raise STRlingError(message)

        clean_patterns.append(pattern)

    if any(p.composite for p in clean_patterns):
        message = """
        Method: simply.not_in_chars(*patterns)

        All patterns must be non-composite.
        """
        raise STRlingError(message)

    joined = r''
    for pattern in clean_patterns:
        if len(str(pattern)) > 1 and str(pattern)[-1] == '}' and str(pattern)[-2] != "\\":
            message = """
            Method: simply.not_in_chars(*patterns)

            Patterns must not have specified ranges.
            """
            raise STRlingError(message)

        if pattern.custom_set:
            if pattern.negated:
                joined += str(pattern)[2:-1]  # [^chars] => chars
            else:
                joined += str(pattern)[1:-1]  # [chars] => chars
        else:
            joined += str(pattern)

    new_pattern = f'[^{joined}]'
    return Pattern(new_pattern, custom_set=True, negated=True)

File: bindings/java/src/main/java/com/strling/simply/Sets.java
package com.strling.simply;

import com.strling.core.Nodes.*;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;

/**
 * Character set and range functions for pattern matching in STRling.
 *
 * <p>This module provides functions for creating character class patterns, including
 * ranges (between), custom sets (customSet), and utilities for combining sets.
 * Character sets are fundamental building blocks for matching specific groups of
 * characters, and these functions make it easy to define complex character matching
 * rules without dealing with raw regex character class syntax.</p>
 */
public class Sets {
    
    /**
     * Matches all characters within and including the start and end of a letter or digit range.
     *
     * @param start The starting character or digit of the range
     * @param end The ending character or digit of the range
     * @param minRep The minimum number of characters to match
     * @param maxRep The maximum number of characters to match
     * @return An instance of the Pattern class
     */
    public static Pattern between(Object start, Object end, Integer minRep, Integer maxRep) {
        if (start instanceof String && end instanceof String) {
            return betweenChars((String) start, (String) end, minRep, maxRep);
        } else if (start instanceof Integer && end instanceof Integer) {
            return betweenDigits((Integer) start, (Integer) end, minRep, maxRep);
        } else {
            String message = "\n" +
                "Method: simply.between(start, end)\n\n" +
                "The 'start' and 'end' arguments must both be integers (0-9) or letters of the same case (A-Z or a-z).";
            throw new STRlingError(message);
        }
    }
    
    /**
     * Matches all characters within and including the start and end of a letter or digit range.
     *
     * @param start The starting character or digit of the range
     * @param end The ending character or digit of the range
     * @return An instance of the Pattern class
     */
    public static Pattern between(Object start, Object end) {
        return between(start, end, null, null);
    }
    
    /**
     * Helper method for between with character ranges.
     */
    private static Pattern betweenChars(String start, String end, Integer minRep, Integer maxRep) {
        // Validation logic
        if (start.length() != 1 || end.length() != 1) {
            throw new STRlingError("The 'start' and 'end' characters must be single letters.");
        }
        
        if (!start.matches("[a-zA-Z]") || !end.matches("[a-zA-Z]")) {
            throw new STRlingError("The 'start' and 'end' must be alphabetical characters.");
        }
        
        boolean startLower = Character.isLowerCase(start.charAt(0));
        boolean endLower = Character.isLowerCase(end.charAt(0));
        if (startLower != endLower) {
            throw new STRlingError("The 'start' and 'end' characters must be of the same case.");
        }
        
        if (start.charAt(0) > end.charAt(0)) {
            throw new STRlingError("The 'start' character must not be lexicographically greater than the 'end' character.");
        }
        
        ClassRange rangeNode = new ClassRange(start, end);
        Node node = new CharClass(false, Arrays.asList(rangeNode));
        Pattern p = new Pattern(node, true, false, false);
        return (minRep != null) ? p.call(minRep, maxRep) : p;
    }
    
    /**
     * Helper method for between with digit ranges.
     */
    private static Pattern betweenDigits(Integer start, Integer end, Integer minRep, Integer maxRep) {
        // Validation logic
        if (start > end) {
            throw new STRlingError("The 'start' integer must not be greater than the 'end' integer.");
        }
        
        if (start < 0 || start > 9 || end < 0 || end > 9) {
            throw new STRlingError("The 'start' and 'end' integers must be single digits (0-9).");
        }
        
        ClassRange rangeNode = new ClassRange(String.valueOf(start), String.valueOf(end));
        Node node = new CharClass(false, Arrays.asList(rangeNode));
        Pattern p = new Pattern(node, true, false, false);
        return (minRep != null) ? p.call(minRep, maxRep) : p;
    }
    
    /**
     * Matches any character not within or including the start and end of a letter or digit range.
     *
     * @param start The starting character or digit of the range
     * @param end The ending character or digit of the range
     * @param minRep The minimum number of characters to match
     * @param maxRep The maximum number of characters to match
     * @return An instance of the Pattern class
     */
    public static Pattern notBetween(Object start, Object end, Integer minRep, Integer maxRep) {
        if (start instanceof String && end instanceof String) {
            return notBetweenChars((String) start, (String) end, minRep, maxRep);
        } else if (start instanceof Integer && end instanceof Integer) {
            return notBetweenDigits((Integer) start, (Integer) end, minRep, maxRep);
        } else {
            String message = "\n" +
                "Method: simply.notBetween(start, end)\n\n" +
                "The 'start' and 'end' arguments must both be integers (0-9) or letters of the same case (A-Z or a-z).";
            throw new STRlingError(message);
        }
    }
    
    /**
     * Matches any character not within or including the start and end of a letter or digit range.
     *
     * @param start The starting character or digit of the range
     * @param end The ending character or digit of the range
     * @return An instance of the Pattern class
     */
    public static Pattern notBetween(Object start, Object end) {
        return notBetween(start, end, null, null);
    }
    
    /**
     * Helper method for notBetween with character ranges.
     */
    private static Pattern notBetweenChars(String start, String end, Integer minRep, Integer maxRep) {
        // Validation logic (similar to betweenChars)
        if (start.length() != 1 || end.length() != 1) {
            throw new STRlingError("The 'start' and 'end' characters must be single letters.");
        }
        
        if (!start.matches("[a-zA-Z]") || !end.matches("[a-zA-Z]")) {
            throw new STRlingError("The 'start' and 'end' must be alphabetical characters.");
        }
        
        boolean startLower = Character.isLowerCase(start.charAt(0));
        boolean endLower = Character.isLowerCase(end.charAt(0));
        if (startLower != endLower) {
            throw new STRlingError("The 'start' and 'end' characters must be of the same case.");
        }
        
        if (start.charAt(0) > end.charAt(0)) {
            throw new STRlingError("The 'start' character must not be lexicographically greater than the 'end' character.");
        }
        
        ClassRange rangeNode = new ClassRange(start, end);
        Node node = new CharClass(true, Arrays.asList(rangeNode));  // negated = true
        Pattern p = new Pattern(node, true, false, false);
        return (minRep != null) ? p.call(minRep, maxRep) : p;
    }
    
    /**
     * Helper method for notBetween with digit ranges.
     */
    private static Pattern notBetweenDigits(Integer start, Integer end, Integer minRep, Integer maxRep) {
        // Validation logic
        if (start > end) {
            throw new STRlingError("The 'start' integer must not be greater than the 'end' integer.");
        }
        
        if (start < 0 || start > 9 || end < 0 || end > 9) {
            throw new STRlingError("The 'start' and 'end' integers must be single digits (0-9).");
        }
        
        ClassRange rangeNode = new ClassRange(String.valueOf(start), String.valueOf(end));
        Node node = new CharClass(true, Arrays.asList(rangeNode));  // negated = true
        Pattern p = new Pattern(node, true, false, false);
        return (minRep != null) ? p.call(minRep, maxRep) : p;
    }
    
    /**
     * Matches any of the provided patterns, but they can't include subpatterns.
     *
     * @param patterns One or more non-composite patterns to match
     * @return An instance of the Pattern class
     */
    public static Pattern inChars(Object... patterns) {
        List<Pattern> cleanPatterns = new ArrayList<>();
        for (Object obj : patterns) {
            if (obj instanceof String) {
                obj = Pattern.lit((String) obj);
            }
            
            if (!(obj instanceof Pattern)) {
                String message = "\n" +
                    "Method: simply.inChars(...patterns)\n\n" +
                    "The parameters must be instances of Pattern or string.\n\n" +
                    "Use a string such as \"123abc$\" to match literal characters, or use a predefined set like simply.letter().";
                throw new STRlingError(message);
            }
            
            cleanPatterns.add((Pattern) obj);
        }
        
        // Check if any pattern is composite (Seq, Alt, Group, Quant, Look, etc.)
        String[] compositeNodeTypes = {"Seq", "Alt", "Group", "Quant", "Look"};
        for (Pattern p : cleanPatterns) {
            String nodeType = p.node.getClass().getSimpleName();
            for (String compType : compositeNodeTypes) {
                if (nodeType.equals(compType)) {
                    String message = "\n" +
                        "Method: simply.inChars(...patterns)\n\n" +
                        "All patterns must be non-composite.";
                    throw new STRlingError(message);
                }
            }
        }
        
        // Build the character class items by extracting them from input patterns
        List<ClassItem> items = new ArrayList<>();
        for (Pattern pattern : cleanPatterns) {
            String nodeType = pattern.node.getClass().getSimpleName();
            if (nodeType.equals("Lit")) {
                // For literals, add each character as a ClassLiteral item
                Lit litNode = (Lit) pattern.node;
                for (char c : litNode.value.toCharArray()) {
                    items.add(new ClassLiteral(String.valueOf(c)));
                }
            } else if (nodeType.equals("CharClass")) {
                // For character classes, add their items directly
                CharClass ccNode = (CharClass) pattern.node;
                items.addAll(ccNode.items);
            }
            // Handle other node types as needed
        }
        
        Node node = new CharClass(false, items);
        return new Pattern(node, true, false, false);
    }
    
    /**
     * Matches anything but the provided patterns, but they can't include subpatterns.
     *
     * @param patterns One or more non-composite patterns to avoid
     * @return An instance of the Pattern class
     */
    public static Pattern notInChars(Object... patterns) {
        List<Pattern> cleanPatterns = new ArrayList<>();
        for (Object obj : patterns) {
            if (obj instanceof String) {
                obj = Pattern.lit((String) obj);
            }
            
            if (!(obj instanceof Pattern)) {
                String message = "\n" +
                    "Method: simply.notInChars(...patterns)\n\n" +
                    "The parameters must be instances of Pattern or string.\n\n" +
                    "Use a string such as \"123abc$\" to match literal characters, or use a predefined set like simply.letter().";
                throw new STRlingError(message);
            }
            
            cleanPatterns.add((Pattern) obj);
        }
        
        // Check if any pattern is composite
        for (Pattern p : cleanPatterns) {
            if (p.isComposite()) {
                String message = "\n" +
                    "Method: simply.notInChars(...patterns)\n\n" +
                    "All patterns must be non-composite.";
                throw new STRlingError(message);
            }
        }
        
        // Build the character class items by extracting them from input patterns
        List<ClassItem> items = new ArrayList<>();
        for (Pattern pattern : cleanPatterns) {
            String nodeType = pattern.node.getClass().getSimpleName();
            if (nodeType.equals("Lit")) {
                // For literals, add each character as a ClassLiteral item
                Lit litNode = (Lit) pattern.node;
                for (char c : litNode.value.toCharArray()) {
                    items.add(new ClassLiteral(String.valueOf(c)));
                }
            } else if (nodeType.equals("CharClass")) {
                // For character classes, add their items directly
                CharClass ccNode = (CharClass) pattern.node;
                items.addAll(ccNode.items);
            }
            // Handle other node types as needed
        }
        
        Node node = new CharClass(true, items);
        return new Pattern(node, true, false, false);
    }
}
